AB#3462686 Implemented Common logic to retrieve broker latency duration from result Broker Bundle, Fixes AB#3462686

Prvnkmr337 opened the pull request:
Fixes: [AB#3462686](https://identitydivision.visualstudio.com/fac9d424-53d2-45c0-91b5-ef6ba7a6bf26/_workitems/edit/3462686) Implemented Common logic to retrieve broker latency duration from result Broker Bundle

- Added BrokerPerformanceMetrics.java to hold broker request latency data.
- AcquireTokenResult and BaseException hold instance of BrokerPerformanceMetrics for the related Broker request and response.
- MsalBrokerResultAdapter updated to read BrokerPerformanceMetrics data from Broker result bundle to be populated in AcquireTokenResult and BaseException.

Commit:
Adds BrokerPerformanceMetrics class and integrate performance metrics into token acquisition flow.

- Added BrokerPerformanceMetrics.java to hold broker request latency data.
- AcquireTokenResult and BaseException hold instance of BrokerPerformanceMetrics for the related Broker request and response.
- MsalBrokerResultAdapter updated to read BrokerPerformanceMetrics data from Broker result bundle to be populated in AcquireTokenResult and BaseException.

Commit:
Update:
1. Updated change log with details.

Commit:
Adds BrokerPerformanceMetrics class and integrate performance metrics into token acquisition flow.

- Added BrokerPerformanceMetrics.java to hold broker request latency data.
- AcquireTokenResult and BaseException hold instance of BrokerPerformanceMetrics for the related Broker request and response.
- MsalBrokerResultAdapter updated to read BrokerPerformanceMetrics data from Broker result bundle to be populated in AcquireTokenResult and BaseException.

Updates:
1. Addressed CR comments.

Prvnkmr337 commented on common4j/src/main/com/microsoft/identity/common/java/broker/BrokerPerformanceMetrics.java:
```diff
@@ -0,0 +1,70 @@
+        this.brokerResponseGenerationTimestamp = brokerResponseGenerationTimestamp;
+        this.brokerHandlingTime = brokerResponseGenerationTimestamp - brokerRequestReceivedTimestamp;
+        this.responseLatency = System.currentTimeMillis() - brokerResponseGenerationTimestamp;
+    }
```
Updated to handle the negative scenario.

mohitc1 commented:
> public class BaseException extends Exception implements IErrorInformation, ITelemetryAccessor {

Add an interface something like IBrokerPerformanceMetricsProvider with method getBrokerPerformanceMetrics() and make BaseException and AcquireTokenResult implement it.

This is similar to suggestion Nick had in our meeting.

---
Refers to: common4j/src/main/com/microsoft/identity/common/java/exception/BaseException.java:42 in 2555cdb. [](commit_id = 2555cdb742087c9351fb1d981e4147a7a5ef729c, deletion_comment = False)

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/BrokerPerformanceMetrics.java:
```diff
@@ -0,0 +1,104 @@
+ * Holds broker performance metrics derived from timestamps in authentication flows.
+ * Calculates processing and latency durations at construction time.
+ */
+public class BrokerPerformanceMetrics {
```
>BrokerPerformanceMetrics [](http://example.com/codeflow?start=13&length=24)

Suggest using Kotlin for new classes

Commit:
Addressed comments, including IBrokerPerformanceMetricsProvider and creating as kotlin classes

Commit:
Fixed Nullable related Spotbug failures

Commit:
Fixed Spotbug main failure in MsalBrokerResultAdapter

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/kotlin/broker/BrokerPerformanceMetrics.kt:
```diff
@@ -0,0 +1,96 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.kotlin.broker
```
>com.microsoft.identity.common.kotlin.broker [](http://example.com/codeflow?start=8&length=43)

You may keep the package previous one. other kotlin files are present there only

Prvnkmr337 replied:
Addressed PR comments, moved new kotlin files to 'common4j/src/main/com/microsoft/identity/common/java/' to keep it consistent with rest of kotlin file location.

Commit:
Update:
1. Address PR comments, moved new kotlin files to 'common4j/src/main/com/microsoft/identity/common/java/' to keep it consistent with rest of kotlin file location.

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/BrokerPerformanceMetrics.java:
```diff
@@ -0,0 +1,70 @@
+package com.microsoft.identity.common.java.broker;
```
@Prvnkmr337

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/BrokerPerformanceMetrics.kt:
```diff
@@ -0,0 +1,74 @@
+package com.microsoft.identity.common.java.broker
```
Missing license statement

Prvnkmr337 replied:
Looks it got removed while moving back to java directory, will fix it.

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/IBrokerPerformanceMetricsProvider.kt:
```diff
@@ -0,0 +1,5 @@
+package com.microsoft.identity.common.java.broker
+
+interface IBrokerPerformanceMetricsProvider {
```
documentation

Prvnkmr337 replied:
Fixed

mohitc1 replied:
I mean documentation for the interface and methods

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/IBrokerPerformanceMetricsProvider.kt:
```diff
@@ -0,0 +1,5 @@
+package com.microsoft.identity.common.java.broker
+
+interface IBrokerPerformanceMetricsProvider {
+    var brokerPerformanceMetrics: BrokerPerformanceMetrics
```
>var [](http://example.com/codeflow?start=4&length=3)

just val? So we can have setter only in implementers.

Prvnkmr337 replied:
No, it needs to be **var** to provide ability to set values. 
Changing to **val** will provide only getter method and not setter method.

mohitc1 replied:
I mean you can have setter method only in implementer classes. For consumer of interface only get is good

Prvnkmr337 replied:
Got it , makes sense. Will update it in following commit.

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/broker/IBrokerPerformanceMetricsProvider.kt:
```diff
@@ -0,0 +1,5 @@
+package com.microsoft.identity.common.java.broker
+
+interface IBrokerPerformanceMetricsProvider {
+    var brokerPerformanceMetrics: BrokerPerformanceMetrics
```
>BrokerPerformanceMetrics [](http://example.com/codeflow?start=34&length=24)

Consider making return type BrokerPerformanceMetrics?, so this can return null.

I feel EMPTY can make caller side process the values erroneously, by considering them real, which can corrupt analysis

Commit:
Update:
1. Addressed PR comments.

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/exception/BaseException.java:
```diff
@@ -216,6 +220,14 @@ public String getUsername() {
+        this.mBrokerPerformanceMetrics = brokerPerformanceMetrics;
+    }
+
+    public BrokerPerformanceMetrics getBrokerPerformanceMetrics() {
```
Override

Commit:
Addressed PR comments

mohitc1 reviewed (approved):
:shipit:

Commit:
Added documentation for IBrokerPerformanceMetricsProvider

siddhijain reviewed (approved):
LGTM

Prvnkmr337 committed:
Merge branch 'dev' into prsaminathan/read-latency-duration
